File: Itido-BMS/RentAPI/Rent/Models/TimePlanning/WorkContract.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using Rent.Helpers;
using Rent.Models.Projects;

namespace Rent.Models.TimePlanning
{

    public class WorkContract
    {
        public int ID { get; set; }
        public int? ContractID { get; set; }

        public string Note { get; set; }
        public DateTime FromDate { get; set; }
        public DateTime ToDate { get; set; }
        public bool IsVisible { get; set; }

        public virtual Contract Contract { get; set; }


        public virtual ICollection<WorkHoliday> WorkHolidays { get; set; }
        public virtual ICollection<WorkDay> WorkDays { get; set; }

        public int? ProjectItemID { get; set; }
        public virtual ProjectItem ProjectItem { get; set; }

        public static Expression<Func<WorkContract, dynamic>> BasicDTO()
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.ToDate,
                v.FromDate,
                v.IsVisible,
            } : null;
        }

        public static Expression<Func<WorkContract, dynamic>> StandardDTO()
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.ToDate,
                v.FromDate,
                v.IsVisible,

                // project 
                ProjectItem = ProjectItem.BasicDTO().Compile()(v.ProjectItem),
                ProjectItem_project = v.ProjectItem != null ? Project.BasicDTO().Compile()(v.ProjectItem.Project) : null,

                // contract
                Contract = Contract.BasicDTO().Compile()(v.Contract),
                Contract_user = v.Contract != null ? User.BasicDTO().Compile()(v.Contract.User) : null,
            } : null;
        }

        public static Expression<Func<WorkContract, dynamic>> DetailedDTO()
        {
            return v => v != null ?
            new
            {
                v.ID,
                v.Note,
                v.ToDate,
                v.FromDate,
                v.IsVisible,

                // project 
                ProjectItem = ProjectItem.BasicDTO().Compile()(v.ProjectItem),
                ProjectItem_project = v.ProjectItem != null ? Project.BasicDTO().Compile()(v.ProjectItem.Project) : null,

                // contract
                Contract = Contract.BasicDTO().Compile()(v.Contract),
                Contract_user = v.Contract != null ? User.BasicDTO().Compile()(v.Contract.User) : null,

                v.WorkDays,
                v.WorkHolidays,
            } : null;
        }
    }
}

File: Itido-BMS/dart/bms_dart/lib/src/models/work_contract.dart
import 'package:bms_dart/models.dart';

import 'date_time_converter.dart';

class WorkContract {
  int id;
  String note;
  DateTime fromDate;
  DateTime toDate;
  bool isVisible;

  Contract contract;
  Project project;

  List<WorkDay> workDays;
  List<WorkHoliday> workHolidays;

  WorkContract({
    // this.contractId,
    // this._projectId,
    this.note,
    this.fromDate,
    this.toDate,
    this.isVisible,
    this.workDays,
    this.workHolidays,
  });

  WorkContract.fromJson(json)
      : this.id = json['id'],
        this.note = json['note'],
        this.fromDate = toDateTime(json['fromDate']),
        this.toDate = toDateTime(json['toDate']),
        this.isVisible = json['isVisible'],
        this.project = Project.fromJson(json['projectItem_project']),
        this.contract = Contract.fromJson(
          json['contract'],
          rootJson: json,
          path: 'contract',
        ),
        this.workDays = List.castFrom(json['workDays'] ?? [])
            .map((j) => WorkDay.fromJson(j))
            .toList(),
        this.workHolidays = List.castFrom(json['workHolidays'] ?? [])
            .map((j) => WorkHoliday.fromJson(j))
            .toList();

  Map<String, dynamic> toMap() => {
        'id': this.id,
        'note': this.note,
        'fromDate': this.fromDate.toString(),
        'toDate': this.toDate.toString(),
        'isVisible': this.isVisible,
        'workDays': this.workDays?.map((w) => w.toMap())?.toList(),
        'workHolidays': this.workHolidays?.map((w) => w.toMap())?.toList(),
      };

  bool get isOwned => contract != null;

  User get user => isOwned && contract.user != null ? contract.user : null;
}
